### PR TITLE
Add metrics middleware

### PR DESCRIPTION
**Pull request by fredericbarthelet**
What does this implement/fix? Explain your changes.
---------------------------------------------------
Implements #618 

Where has this been tested?
---------------------------
**Node.js Versions:** 12

**Middy Versions:** 2.0.0-alpha6

**AWS SDK Versions:** No relevant

Todo list
---------

[ x ] Feature/Fix fully implemented
[ x ] Added tests
[ x ] Updated relevant documentation
[ x ] Updated relevant examples

### CODE REVIEW

**Comment by willfarrell** [comment]
I like it, I think this would be a great addition. Not 100% sold on the name yet. What do you think of `cloudwatch-metrics`? It feels long, but `metrics` feels too generic. @lmammino have any opinions on the name?

Would you be able to update this for merging into `release/2.x`? I'd like to avoid new additions to v1.x at this point.

Skimming through it's npm docs is `setProperty("RequestId", handler.path.to.RequestId)` required? Or is it set by default and this if for overriding?

**Comment by lmammino** [comment]
I like `cloudwatch-metrics`. Before looking at the code, I was actually thinking that this middleware was providing generic hooks to send metrics to an arbitrary backend like statsd.

**Comment by fredericbarthelet** [comment]
Thanks @willfarrell and @lmammino :)
I implemented your feedbacks :
- rebase on release/2.x 
- rename to `cloudwatch-metrics`

> Skimming through it's npm docs is setProperty("RequestId", handler.path.to.RequestId) required? Or is it set by default and this if for overriding?

`setProperty`, like `putMetric`, are part of the API available on the new `context.metrics` property. This method allows adding context specific property that you'd like to see on you Cloudwatch Insights console. There are no defaults for it, unlike namespace  and dimensions.

For exemple, in a lambda processing payment, you could use `setProperty("Total VAT", 89)` in case you'd like to have this metric available. Maybe my exemple (taken directly from the libs documentation) is not that much clear  ?

File: packages/cloudwatch-metrics/package.json [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    "url": "https://github.com/middyjs/middy/issues"
+  },
+  "homepage": "https://github.com/middyjs/middy#readme",
+  "dependencies": {
+    "aws-embedded-metrics": "^2.0.4"
+  },
+  "devDependencies": {
+    "@middy/core": "^1.5.2",
+    "es6-promisify": "^6.0.2"
+  },
```
Comment (willfarrell) [inline packages/cloudwatch-metrics/package.json:58]: Can you remove devDependencies, they are not needed in v2

File: packages/cloudwatch-metrics/package.json [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    "type": "git",
+    "url": "github:middyjs/middy",
+    "directory": "packages/cloudwatch-metrics"
+  },
+  "bugs": {
+    "url": "https://github.com/middyjs/middy/issues"
+  },
+  "homepage": "https://github.com/middyjs/middy#readme",
+  "dependencies": {
+    "aws-embedded-metrics": "^2.0.4"
```
Comment (willfarrell) [inline packages/cloudwatch-metrics/package.json:53]: Can we lock the version ` "aws-embedded-metrics": "2.0.4"`.

**Review by willfarrell — CHANGES_REQUESTED** [review]

**Review by willfarrell — APPROVED** [review]

**Comment by willfarrell** [comment]
@fredericbarthelet can you commit the `package-lock.json` as well please. It's required for the tests.

**Comment by fredericbarthelet** [comment]
@willfarrell `package-lock.json` added :)
I noticed all package-lock uses `lockfileVersion: 1` except root package-lock, using v2 (from npm 7). Is this expected ? Should I target v1 or v2 for my lock file ?

https://github.com/middyjs/middy/blob/b75acc929df0f4f1bc75b99d6b9404b83b1e30af/package-lock.json#L4

**Comment by willfarrell** [comment]
The lockfileVersion change is unexpected but good to know. My local must have finally updated causing the change. Can you update to use v1. The CI uses npm v6.

**Comment by fredericbarthelet** [comment]
@willfarrell, I added a commit to change main `package-lock.json` file back to lockFileVersion 1 :)

**Comment by willfarrell** [comment]
Thanks, looks like it's still failing on other lockfile v2's. I pushed a fix for these. One last rebase should do the trick, then we can merge in. Thank you for your patience on resolving this lockfile version issue. We have another PR that will properly bump all to lockfile v2 in the works.

**Comment by fredericbarthelet** [comment]
@willfarrell I just rebased, everything's green now :)

**Comment by willfarrell** [comment]
Great! I'm just waiting on another PR before minting a new version. Hope that's okay.

**Comment by fredericbarthelet** [comment]
Thanks @willfarrell :) No problem waiting a bit more for v2 ;)

**Comment by willfarrell** [comment]
@fredericbarthelet With AWS powertools released (https://github.com/awslabs/aws-lambda-powertools-typescript/tree/main/packages/metrics). Is their implementation better/easier to use than this middleware?

**Comment by fredericbarthelet** [comment]
Hi @willfarrell, ultimately the powertool implementation is more complete and will be the prevailing implementation. However, in its current state, the powertool remains a beta-only, not to be used in production solution. I'd wait out of beta before deprecating the metrics middleware. Once they're out of beta, I'd be happy to update documentation with a migration path for the new solution.

**Comment by willfarrell** [comment]
Perfect, I was thinking the same but wanted to make sure I got your feedback as you championed this middleware. I've added to the v4 list to revisit then, unless power tools hits production ready before v3 is released.

**Pull request merged**